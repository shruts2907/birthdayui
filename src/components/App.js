import React, { Component } from "react";
import Fetch from "./FetchData";
import Sort from "./Sort";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      data: [],
      sortOrder: "name",
      sortObjOrder: "first",
      sortDir: "asc",
	  queryText: ''
    };


    this.searchApts = this.searchApts.bind(this);
    this.changeOrder = this.changeOrder.bind(this);
    this.filtereList = this.filteredList.bind(this);
    this.url = "https://next.json-generator.com/api/json/get/E1LpMX5Su";
  }


	/*
		Search the table with query passed
		@param query string
	*/
	searchApts(query) {
		this.setState({ queryText: query });
	}
	
	/*
		Captures Sort order event and updates state
		@param order string
		@param dir string
		@param name string
	*/
	changeOrder(order, dir, name) {
		this.setState({
		  sortOrder: order,
		  sortObjOrder: name,
		  sortDir: dir
		});
	}
	/*
		Fitler the json data with response to sort order change
	*/  
	filteredList() {
		let order;
		let filtereList = this.state.data;

		if (this.state.sortDir === "asc") {
		  order = 1;
		} else {
		  order = -1;
		}

		filtereList = filtereList.sort((a, b) => {
		  if (
			typeof a[this.state.sortOrder] === "object" ||
			typeof b[this.state.sortOrder] === "object"
		  ) {
			if (
			  a[this.state.sortOrder][this.state.sortObjOrder].toLowerCase() <
			  b[this.state.sortOrder][this.state.sortObjOrder].toLowerCase()
			) {
			  return -1 * order;
			} else {
			  return 1 * order;
			}
		  } else {
			if (
			  a[this.state.sortOrder].toLowerCase() <
			  b[this.state.sortOrder].toLowerCase()
			) {
			  return -1 * order;
			} else {
			  return 1 * order;
			}
		  }
		})
		.filter(eachItem => {
			return (
			  eachItem['name']['first']
				.toLowerCase()
				.includes(this.state.queryText.toLowerCase()) ||
			  eachItem['name']['last']
				.toLowerCase()
				.includes(this.state.queryText.toLowerCase()) ||
			  eachItem['country']
				.toLowerCase()
				.includes(this.state.queryText.toLowerCase())
			);
		  });
		  
		return filtereList;
	}
  
	componentDidMount() {
		fetch(this.url)
			.then(response => response.json())
			.then(result => {
				const list = result.map(item => {
					return item;
				});
			
			this.setState({
				data: list
			});
		}, (error) => {
				if (error) {
				  console.log("Error while fetching data: "+error);
				}
			}
		);
	}
	render() {

	let filtereList = this.filteredList();

		return (
		  <div className="App wrapper">
			<h3 className="greetings">Hello, Check Who's Birthday is Coming Up!</h3>
			<Sort
			  orderBy={this.state.sortOrder}
			  orderDir={this.state.sortDir}
			  objOrder={this.state.sortObjOrder}
			  changeOrder={this.changeOrder}
			  searchApts={this.searchApts}
			/>
			<Fetch data={filtereList} />
		  </div>
		);
	}
}
export default App;
