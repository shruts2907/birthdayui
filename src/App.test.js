import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from './components/App';

test('Renders greetings Hello', () => {
  const { getByText } = render(<App />);
  const greetElement = getByText(/Hello/i);
  expect(greetElement).toBeInTheDocument();
});

test('App snapshot test updated', () => {
  const component = renderer.create(<App />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
