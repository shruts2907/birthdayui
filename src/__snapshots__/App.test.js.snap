// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot test updated 1`] = `
<div
  className="App wrapper"
>
  <h3
    className="greetings"
  >
    Hello, Check Who's Birthday is Coming Up!
  </h3>
  <div
    className="search-appointments row justify-content"
  >
    <div
      className="col-md-6"
    >
      <div
        className="input-group"
      >
        <input
          aria-label="Search List"
          className="form-control"
          id="SearchList"
          onChange={[Function]}
          type="text"
        />
        <div
          className="input-group-append"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            className="btn sort-btn-css dropdown-toggle"
            data-toggle="dropdown"
            type="button"
          >
            Sort by: 
            <span
              className="caret"
            />
          </button>
          <div
            className="sort-menu dropdown-menu dropdown-menu-right"
          >
            <button
              className="sort-by dropdown-item active"
              href="#"
              onClick={[Function]}
            >
              First Name
            </button>
            <button
              className="sort-by dropdown-item "
              href="#"
              onClick={[Function]}
            >
              Last Name
            </button>
            <button
              className="sort-by dropdown-item "
              href="#"
              onClick={[Function]}
            >
              Country
            </button>
            <div
              className="dropdown-divider"
              role="separator"
            />
            <button
              className="sort-by dropdown-item active"
              href="#"
              onClick={[Function]}
            >
              Asc
            </button>
            <button
              className="sort-by dropdown-item "
              href="#"
              onClick={[Function]}
            >
              Desc
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <table>
      <thead>
        <tr>
          <th>
            First Name
          </th>
          <th>
            Last Name
          </th>
          <th>
            Date of Birth
          </th>
          <th>
            Birthday status
          </th>
          <th
            className="country-item"
          >
            Country
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
